db/hints/internal/hint_storage.cc

```diff
@@ -249,36 +242,34 @@ void rebalance_segments(const fs::path& hint_directory, hints_segments_map& segm
 ///                           E is the number of endpoints for which hints were ever created.
 ///
 /// \param hint_directory a root hint directory
-void remove_irrelevant_shards_directories(const fs::path& hint_directory) {
+future<> remove_irrelevant_shards_directories(const fs::path& hint_directory) {
     // Shard level.
-    scan_shard_hint_directories(hint_directory, [] (fs::path dir, directory_entry de, unsigned shard_id) {
+    co_await scan_shard_hint_directories(hint_directory,
+            [] (fs::path dir, directory_entry de, unsigned shard_id) -> future<> {
         if (shard_id >= smp::count) {
             // IP level.
-            return lister::scan_dir(dir / de.name, lister::dir_entry_types::full(),
+            co_await lister::scan_dir(dir / de.name, lister::dir_entry_types::full(),
                     lister::show_hidden::yes, [] (fs::path dir, directory_entry de) {
                 return io_check(remove_file, (dir / de.name).native());
-            }).then([shard_base_dir = dir, shard_entry = de] {
-                return io_check(remove_file, (shard_base_dir / shard_entry.name).native());
             });
+            
+            co_await io_check(remove_file, (dir / de.name).native());
         }
-
-        return make_ready_future<>();
-    }).get();
+    });
 }
 
 } // anonymous namespace
 
 future<> rebalance(fs::path hints_directory) {
-    return seastar::async([hints_directory = std::move(hints_directory)] {
-        // Scan currently present hints segments.
-        hints_segments_map current_hints_segments = get_current_hints_segments(hints_directory);
+    // Scan currently present hint segments.
+    hints_segments_map current_hints_segments = co_await get_current_hints_segments(hints_directory);
 
-        // Move segments to achieve an even distribution of files among all present shards.
-        rebalance_segments(hints_directory, current_hints_segments);
+    // Move segments to achieve an even distribution of files among all present shards.
+    co_await rebalance_segments(hints_directory, current_hints_segments);
 
-        // Remove the directories of shards that are not present anymore - they should not have any segments by now
-        remove_irrelevant_shards_directories(hints_directory);
-    });
+    // Remove the directories of shards that are not present anymore.
+    // They should not have any segments by now.
+    co_await remove_irrelevant_shards_directories(hints_directory);
```

Comment:
Nit: indentation changes should usually be done in a separate commit, to improve the readability of diffs.